austation/code/modules/coilgun/_coilgun.dm

```diff
@@ -0,0 +1,99 @@
+	name = "magnetizer"
+	desc = "A machine that glazes inserted objects with neodymium, making the object magnetive"
+	icon_state = "magnet"
+
+/obj/structure/disposalpipe/coilgun/magnetizer/transfer(obj/structure/disposalholder/H)
+	if(H.contents.len) // is there an object in here?
+		update_icon()
+		var/obj/effect/hvp/boolet
+		for(var/atom/movable/AM in H.contents)
+			if(H.contents.len > 1)
+				visible_message("<span class='warning'>\The [src] can't magnetize more than one object at a time!</span>")
+				AM.forceMove(get_turf(src))
+				continue
+			if(AM == boolet)
+				continue
+			else
+				boolet = new(H)
+				boolet.name = AM.name
+				boolet.desc = AM.desc
+				boolet.icon = AM.icon
+				boolet.icon_state = AM.icon_state
+				boolet.p_speed = 1
+				AM.loc = boolet //put the original inserted objected inside the coilgun projectile
+				boolet.apply_special(AM, TRUE)
+			if(isliving(AM))
+				var/mob/living/L = AM
+				L.adjustBruteLoss(10)
+				if(ishuman(L) && !isdead(L))
+					L.Paralyze(amount = 50, ignore_canstun = TRUE)
+					L.emote("scream")
+					boolet.mass = 5
+					sleep(30)
+					continue
+				else
+					boolet.mass = 3
+					continue
+			if(isitem(AM))
+				var/obj/item/I = AM
+				if(I.w_class)
+					boolet.mass = I.w_class
+					playsound(src.loc, 'sound/machines/ping.ogg', 40, 1)
+					continue
+				else
+					qdel(boolet)
+					qdel(I)
+					return
+	else
+		qdel(H)
+
+	return ..()
+
+/obj/structure/disposalpipe/coilgun/cooler
+	name = "passive coilgun cooler"
+	desc = "A densely packed array of radiator fins designed to passively remove heat from a magnetic projectile, slightly slows down the projectile"
+	icon_state = "p_cooler"
+	var/heat_removal = 2.5 // how much heat we will remove from the projectile
+	var/linear_penalty = TRUE // do we multiply or negate speed_penalty from projectile speed?
+	var/speed_penalty = 2 // multiplies/negates projectile speed by this
+	var/hugbox = FALSE // debug and/or admin abuse
+
+/obj/structure/disposalpipe/coilgun/cooler/active
+	name = "active coilgun cooler"
+	desc = "A tube with multiple small, fast fans used for cooling any projectile that passes through it. Much more effective than a passive cooler but slows the projectile down more"
+	icon = 'austation/icons/obj/atmospherics/pipes/disposal.dmi'
+	icon_state = "a_cooler"
+	heat_removal = 3
+	linear_penalty = FALSE
+	speed_penalty = 0.97
+
+/obj/structure/disposalpipe/coilgun/cooler/transfer(obj/structure/disposalholder/H)
+	if(H.contents.len)
+		for(var/atom/movable/AM in H.contents) // run the loop below for every movable that passes through the charger
+			if(istype(AM, /obj/effect/hvp)) // if it's a projectile, continue
+				var/obj/effect/hvp/projectile = AM
+				projectile.p_heat = max(projectile.p_heat - heat_removal, 0) // projectile can't go below zero
+				if(!hugbox)
+					if(linear_penalty)
+						projectile.p_speed -= speed_penalty
+					else
+						projectile.p_speed *= speed_penalty
```

Comment:
personally, I'd make both of these decrease in exponential fashion, just different amounts.
With the ridiculous speeds possible with a lot of power, the difference between losing 3% of your velocity for a 3 unit reduction in heat and losing a *static* 2 units of speed for a 2.5 unit heat reduction (only a 0.5 unit difference) is a pretty clear choice for the passive every time, unless I've misunderstood something.

What I'm saying is, mixing exponential and linear math doesn't give sane results in a lot of cases.

Reply:
I've barely worked on cooling code, this is a draft PR xd. I only really use this for debug atm.

Once the values are balanced, using the exponential cooler will be better than using the linear cooler for lower velocities, the opposite would apply for higher velocities where the speed increase is smaller